src/AspNet.Security.OAuth.SuperOffice/SuperOfficeAuthenticationHandler.cs

```diff
@@ -0,0 +1,159 @@
+﻿/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.IdentityModel.Tokens;
+
+namespace AspNet.Security.OAuth.SuperOffice
+{
+    /// <summary>
+    /// Defines a handler for authentication using SuperOffice.
+    /// </summary>
+    public class SuperOfficeAuthenticationHandler : OAuthHandler<SuperOfficeAuthenticationOptions>
+    {
+        private readonly JwtSecurityTokenHandler _tokenHandler;
+
+        public SuperOfficeAuthenticationHandler(
+            [NotNull] IOptionsMonitor<SuperOfficeAuthenticationOptions> options,
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock,
+            [NotNull] JwtSecurityTokenHandler tokenHandler)
+            : base(options, logger, encoder, clock)
+        {
+            _tokenHandler = tokenHandler;
+        }
+
+        /// <inheritdoc />
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(
+            [NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties,
+            [NotNull] OAuthTokenResponse tokens)
+        {
+            // the access token is a reference token prefixed with the tenant context identifier
```

Comment:
```suggestion
            // The access token is a reference token prefixed with the tenant context identifier
```

Reply:
@AnthonyYates BTW, access tokens are meant to be completely opaque to clients and parsing them client-side is just terrible practice. Is there really no other way to retrieve the "context id"?

Reply:
Yes, unfortunately the team responsible for the server implementation had their own ideas. I like this feedback though. It has made me rethink the approach and I have refactored this. The same identifier is accessible from the id_token claims, and so I have rewritten the way that is handled in the Handler.